embeddings/task-type-embedding.ipynb

```diff
@@ -0,0 +1,3079 @@
+    "\n",
+    "In this tutorial, we will use 1K rows to measure the search quality. By increasing the rows to 10K you will see better results, but it would take about 30 minutes to finish the tutorial."
+   ],
+   "metadata": {
+    "id": "QYOuS6dbI3Kz"
+   },
+   "id": "QYOuS6dbI3Kz"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## Calculating Q&A search quarity with SEMANTIC_SIMILARITY"
+   ],
+   "metadata": {
+    "id": "JfArfckVWyVX"
+   },
+   "id": "JfArfckVWyVX"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "First, let's test the task type SEMANTIC_SIMILARITY as a baseline.\n",
+    "\n"
+   ],
+   "metadata": {
+    "id": "QjqKPlX9JeXH"
+   },
+   "id": "QjqKPlX9JeXH"
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "### Generate embeddings for questions and answers\n",
+    "\n",
+    "Here we define a wrapper function `get_embeddings()` to make it easier to call the Embeddings API. To specify task type on generating embedding, you can use `TextEmbeddingInput(text, task_type)` function."
+   ],
+   "metadata": {
+    "id": "yEeIoPNKzLad"
+   },
+   "id": "yEeIoPNKzLad"
+  },
+  {
+   "cell_type": "code",
+   "source": [
+    "from google.cloud import aiplatform\n",
+    "from vertexai.language_models import TextEmbeddingInput, TextEmbeddingModel\n",
+    "\n",
+    "# init the aiplatform package\n",
+    "aiplatform.init(project=PROJECT_ID, location=LOCATION)\n",
+    "\n",
+    "MODEL_NAME = \"text-embedding-004\"\n",
+    "\n",
+    "\n",
+    "# a wrapper to get text embeddings for a list of texts\n",
+    "def get_embeddings(texts, task_type):\n",
+    "    model = TextEmbeddingModel.from_pretrained(MODEL_NAME)\n",
+    "    inputs = [TextEmbeddingInput(text, task_type) for text in texts]\n",
+    "    embs = model.get_embeddings(inputs)\n",
+    "    return [emb.values for emb in embs]"
+   ],
+   "metadata": {
+    "id": "mNTyxY0dJsfc",
+    "executionInfo": {
+     "status": "ok",
+     "timestamp": 1726043121125,
+     "user_tz": -540,
+     "elapsed": 3420,
+     "user": {
+      "displayName": "",
+      "userId": ""
+     }
+    }
+   },
+   "id": "mNTyxY0dJsfc",
+   "execution_count": 5,
+   "outputs": []
+  },
+  {
+   "cell_type": "markdown",
+   "source": [
+    "Using this `get_embeddings()` function, we define another function `generate_embs()` that takes a DataFrame and generate embeddings for all rows, with the task types specified for questions and answers. The embeddings will be stored on new columns `q_emb` and `a_emb`."
+   ],
+   "metadata": {
+    "id": "LkxkofhhMUJA"
+   },
+   "id": "LkxkofhhMUJA"
+  },
+  {
+   "cell_type": "code",
+   "source": [
+    "from tqdm import tqdm  # to show the progress bar\n",
+    "\n",
+    "\n",
+    "# generate text embeddings for questions and answers\n",
+    "def generate_embs(df, question_task_type, answer_task_type):\n",
+    "    q_embs = []\n",
+    "    a_embs = []\n",
```

Comment:
Is it efficient to create these separate lists for q&a embs instead of loading directly into the dataframe?

Reply:
I wanted to show the whole process as deltas to the original df, rather than letting the reader following additional separate lists. E.g. Q&As, then adding embs, adding query_ranks... so on. Does it makes sense?

Reply:
That makes sense